File: sdk/python/src/ufds_sdk/webhooks/signature.py
from __future__ import annotations

import hashlib
import hmac
from typing import Optional, Union


def sign_payload(secret: str, raw_body: Union[bytes, str]) -> str:
    """Computes the hex-encoded HMAC-SHA256 signature UFDS expects for a webhook body."""
    body = raw_body.encode("utf-8") if isinstance(raw_body, str) else raw_body
    key = secret.encode("utf-8") if isinstance(secret, str) else secret
    return hmac.new(key, body, hashlib.sha256).hexdigest()


def verify_signature(secret: str, raw_body: Union[bytes, str], signature: Optional[str]) -> bool:
    """Verifies the `X-UFDS-Signature` header against the raw (unparsed) request body.

    Accepts either a bare hex digest or a `sha256=<digest>` prefixed value.
    """
    if not signature:
        return False

    provided = signature[len("sha256=") :] if signature.startswith("sha256=") else signature
    expected = sign_payload(secret, raw_body)
    return hmac.compare_digest(expected, provided)

File: sdk/javascript/src/webhooks/signature.js
import { createHmac, timingSafeEqual } from 'node:crypto';

/** Computes the hex-encoded HMAC-SHA256 signature UFDS expects for a webhook body. */
export function signPayload(secret, rawBody) {
  const body = Buffer.isBuffer(rawBody) ? rawBody : Buffer.from(String(rawBody), 'utf8');
  return createHmac('sha256', secret).update(body).digest('hex');
}

/**
 * Verifies the `X-UFDS-Signature` header against the raw (unparsed) request body.
 * Accepts either a bare hex digest or a `sha256=<digest>` prefixed value.
 */
export function verifySignature(secret, rawBody, signature) {
  if (!signature) return false;

  const provided = String(signature).startsWith('sha256=')
    ? String(signature).slice('sha256='.length)
    : String(signature);

  const expected = signPayload(secret, rawBody);
  const expectedBuf = Buffer.from(expected, 'utf8');
  const providedBuf = Buffer.from(provided, 'utf8');

  if (expectedBuf.length !== providedBuf.length) return false;
  return timingSafeEqual(expectedBuf, providedBuf);
}
